hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationFuture.java

```diff
@@ -380,11 +367,21 @@ private Object resolveApplicationResponse(Object unresolvedResponse) {
         }
 
         if (response instanceof Throwable) {
-            Throwable throwable = ((Throwable) response);
-            if (invocation.remote) {
-                fixRemoteStackTrace((Throwable) response, Thread.currentThread().getStackTrace());
+            try {
+                Throwable throwable = ((Throwable) response);
+                SerializationService serializationService = operationService.serializationService;
+                // Exception needs cloning since stacktrace is modified, otherwise same exception shared between threads.
+                Throwable clonedThrowable = serializationService.toObject(serializationService.toData(throwable));
+                fixRemoteStackTrace(clonedThrowable, Thread.currentThread().getStackTrace());
+                return clonedThrowable;
+            } catch (HazelcastSerializationException e) {
+                // It can happen that the exceptions fails cloning, in that case return the original unmodified exception
+                operationService.logger.warning("Failed to clone exception", e);
+                return response;
+            } catch (HazelcastInstanceNotActiveException e) {
+                // In case of a shutdown, lets return the HazelcastInstanceNotActiveException
```

Comment:
On purpose. Because the stacktrace can't be modified; so the original exception would come out of thin air.